source/scpd/scp/BallotProtocol.d

```diff
@@ -0,0 +1,273 @@
+    //                             const ref Interval check,
+    //                             const ref SCPStatement st);
+
+    // attempts to update p to ballot (updating p' if needed)
+    bool setPrepared(const ref SCPBallot ballot);
+
+    // ** Helper methods to compare two ballots
+
+    // ballot comparison (ordering)
+    // static int compareBallots(std::unique_ptr<SCPBallot> const& b1,
+    //                           std::unique_ptr<SCPBallot> const& b2);
+    static int compareBallots(const ref SCPBallot b1, const ref SCPBallot b2);
+
+    // b1 ~ b2
+    static bool areBallotsCompatible(const ref SCPBallot b1, const ref SCPBallot b2);
+
+    // b1 <= b2 && b1 !~ b2
+    static bool areBallotsLessAndIncompatible(const ref SCPBallot b1,
+                                              const ref SCPBallot b2);
+    // b1 <= b2 && b1 ~ b2
+    static bool areBallotsLessAndCompatible(const ref SCPBallot b1,
+                                            const ref SCPBallot b2);
+
+    // ** statement helper functions
+
+    // returns true if the statement is newer than the one we know about
+    // for a given node.
+    bool isNewerStatement(const ref NodeID nodeID, const ref SCPStatement st);
+
+    // returns true if st is newer than oldst
+    static bool isNewerStatement(const ref SCPStatement oldst,
+                                 const ref SCPStatement st);
+
+    // basic sanity check on statement
+    bool isStatementSane(const ref SCPStatement st, bool self);
+
+    // records the statement in the state machine
+    void recordEnvelope(const ref SCPEnvelope env);
+
+    // ** State related methods
+
+    // helper function that updates the current ballot
+    // this is the lowest level method to update the current ballot and as
+    // such doesn't do any validation
+    // check: verifies that ballot is greater than old one
+    void bumpToBallot(const ref SCPBallot ballot, bool check);
+
+    // switch the local node to the given ballot's value
+    // with the assumption that the ballot is more recent than the one
+    // we have.
+    bool updateCurrentValue(const ref SCPBallot ballot);
+
+    // emits a statement reflecting the nodes' current state
+    // and attempts to make progress
+    void emitCurrentStateStatement();
+
+    // verifies that the internal state is consistent
+    void checkInvariants();
+
+    // create a statement of the given type using the local state
+    SCPStatement createStatement(const ref SCPStatementType type);
+
+    // returns a string representing the slot's state
+    // used for log lines
+    //std::string getLocalState() const;
+
+    shared_ptr!LocalNode getLocalNode();
+
+    // bool federatedAccept(StatementPredicate voted, StatementPredicate accepted);
+    // bool federatedRatify(StatementPredicate voted);
+
+    void startBallotProtocolTimer();
+    void stopBallotProtocolTimer();
+    void checkHeardFromQuorum();
+}
+
+static assert(BallotProtocol.sizeof == 136);
```

Comment:
Long term goal: Make this check automatic

Reply:
Do you know about clicking the "..." on the right and clicking on "Reference in new issue"? I'm about to do that.